determined-ai/determined
# chore: allow non-singleton Authentication

## Description

The authentication.Authentication object was written as a singleton
object for the CLI.  Since the CLI never talked to multiple masters or
wrote to multiple masters, that pattern worked great.

However, the upcoming Python SDK will need more flexibility than that.
Rather than duplicate the existing authentication logic, we can refactor
what exists now to be a bit more flexible so that the Python SDK can
reuse what is already written.

Additionally, this PR introduces an explicit Cert object, and configures all
Python SDK API calls to pass explicit Authentication and Cert objects.

## Test Plan

I am new to the cli authentication logic.  I ran all the following tests manually:

- authentication bootstrapping is automatic
- broken tokens in auth.json get cleaned up automatically
- make sure the fallback to "regular determined user" still works.
- change password for yourself -> token update in file
- change password for other user -> no token update in file
- login works
- logout works
- logout unsets active user
- logout removes token
